src/hotspot/share/gc/shared/c2/barrierSetC2.cpp

```diff
@@ -896,6 +896,263 @@ void BarrierSetC2::clone_at_expansion(PhaseMacroExpand* phase, ArrayCopyNode* ac
 
 #undef XTOP
 
+static bool block_has_safepoint(const Block* block, uint from, uint to) {
+  for (uint i = from; i < to; i++) {
+    if (block->get_node(i)->is_MachSafePoint()) {
+      // Safepoint found
+      return true;
+    }
+  }
+
+  // Safepoint not found
+  return false;
+}
+
+static bool block_has_safepoint(const Block* block) {
+  return block_has_safepoint(block, 0, block->number_of_nodes());
+}
+
+static uint block_index(const Block* block, const Node* node) {
+  for (uint j = 0; j < block->number_of_nodes(); ++j) {
+    if (block->get_node(j) == node) {
+      return j;
+    }
+  }
+  ShouldNotReachHere();
+  return 0;
+}
+
+// Look through various node aliases
+static const Node* look_through_node(const Node* node) {
+  while (node != nullptr) {
+    const Node* new_node = node;
+    if (node->is_Mach()) {
+      const MachNode* const node_mach = node->as_Mach();
+      if (node_mach->ideal_Opcode() == Op_CheckCastPP) {
+        new_node = node->in(1);
+      }
+      if (node_mach->is_SpillCopy()) {
+        new_node = node->in(1);
+      }
+    }
+    if (new_node == node || new_node == nullptr) {
+      break;
+    } else {
+      node = new_node;
+    }
+  }
+
+  return node;
+}
+
+// Whether the given offset is undefined.
+static bool is_undefined(intptr_t offset) {
+  return offset == Type::OffsetTop;
+}
+
+// Whether the given offset is unknown.
+static bool is_unknown(intptr_t offset) {
+  return offset == Type::OffsetBot;
+}
```

Comment:
pre-existing: these two predicates would be nice to have elsewhere in C2 code too (i.e. there's quite a lot of these predicates inlined elsewhere).

Reply:
Maybe something for another RFE of course.

Reply:
Thanks, this will make a good starter RFE.